docs/src/references/c8y-configuration-management.md

```diff
@@ -163,3 +275,117 @@ Note that:
+
+For each kind of child device such a specific client has to be implemented
+and installed on the child-device hardware (To be precise, most often on the child device,
+but not necessarily as, for a child device which software cannot be altered,
+one might have to install a protocol adaptor on the main device).
+
+Here is the protocol that have to be implemented by child-device configuration client.
+This protocol covers 3 interactions:
+* Receiving configuration file updates from thin-edge
+* Sending to thin-edge current configuration files
+* Notifying the list of configuration files to thin-edge
+
+### The child device connect the main thin-edge device
+
+From a TCP point of view, the child devices act as clients
+and all the connections to thin-edge are established by the child devices.
+* Thin-edge opens two ports for MQTT and HTTP over the local network.
+  These ports are controlled on the main device `tedge config`:
+  * `mqtt.external.bind_address`
+  * `mqtt.external.port`
+  * `http.external.port`
+* The child devices must know the main device IP address.
+  * This is the address set for `mqtt.external.bind_address` on the main device.
+  * For the very specific case, where the software for the child device run on the main device,
+    this connection address can be the `localhost`.
+* On start, the child-device specific software for configuration management,
+  must connect over MQTT to thin-edge device.
+  * It must use a session name that is unique on the thin-edge bus.
+    `"$CHILD_DEVICE_ID/configuration"` is the recommendation.
+  * It must subscribe to `tedge/$CHILD_DEVICE_ID/commands/req/#` to receive requests.
+  * It has to publish responses under `tedge/$CHILD_DEVICE_ID/commands/res/#`.
+* On demand, the child-device software has to send HTTP requests to thin-edge.
+  * These requests are REST requests (GET/PUT/DELETE) to exchange files.
+  * The urls are forged by thin-edge telling the child devices
+    where to get and put configuration files.
+* Currently, all these connections are done without TLS support.
+  * Hence, the main device must be configured so these MQTT and HTTP ports
+    are only open on a local trusted network.
+
+### The child device downloads configuration updates on notification
+
+When a configuration file for a child device
+is received from the cloud by the `c8y_configuration_plugin` daemon,
+the configuration plugin manages the transfer of this file to the child device
+and notifies the cloud on the progress of this configuration update operation.
+
+1. On reception of a configuration update for a child device named `$CHILD_DEVICE_ID`
+   of a file `{ path = $PATH, type = $TYPE }`defined in `$CHILD_DEVICE_ID/c8y-configuration-plugin.toml` 
+   the `c8y_configuration_plugin` downloads the new configuration content in `/tmp`,
+   and, on success, moves this content under the thin-edge file transfer HTTP root.
+   * The file is moved to `$TEDGE_HTTP_ROOT/config_snapshot/$CHILD_DEVICE_ID/$TYPE`
```

Comment:
Two things.
1. I believe that we use `config_snapshot` for `c8y_UploadConfigFile` (upload a snapshot to c8y), and `config_update` for `c8y_DownloadConfigFile` (deploying a snapshot to device). So, here should be `config_update` instead of `config_snapshot`. The implementation of our prototype and tedge_agent enhancement by @initard is distinguish them like that. (Note: The proposal from @cstoidner is mixed on that point, I guess it's mistake. [here](https://github.com/thin-edge/thin-edge.io/pull/1236/files?short_path=0d43010#diff-0d430102a9a7f4e4431390a85806a990b51d2143dbc97bf596a267c893e1f693))
2. The order of `config_snapshot/$CHILD_DEVICE_ID`. The prototype and tedge_agent has the oppisite order, e.g.
`/var/tedge/file-transfer/<child_device_id>/config_snapshot/<type>`. Which one should be?

```suggestion
   * The file is moved to `$TEDGE_HTTP_ROOT/$CHILD_DEVICE_ID/config_update/$TYPE`
```

Reply:
1. You are correct => fixed
2. There is no reason to change what has been done so far => updated